### ISSUE TITLE
unexpected order of classpath precedence with resource files

### ISSUE THREAD

**Issue by softprops** [bazelbuild/rules_scala#189]
I'm not sure if I should be reporting this in bazel proper or here, but an issue I ran into recently was the classpath ordering of resource files. In sbt, if I have a module `foo` with a resource named `bar.properties` under `src/main/resources` and a test that overrides that under `src/test/resources/bar.properties`, I can be sure that when I run my test, `bar.properties` will be read first when reading from the classloader as a stream. 

However, with `scala_test` rules, if I have a similar declaration of resources in a `scala_library` dep of my `scala_test`, the copy from the `scala_library`'s resources seems to get resolved first. 

Again, I'm not sure if this is a specific problem to these rules or bazel in general. It also may be an antipattern that I shouldn't be doing in the first place :)

**Comment by ittaiz** [comment]
@softprops any chance for a printout of the classpath? you can `cat` the launcher script to see it. Additionally what I'd appreciate is if you could try and reverse the order of the two relevant jars and report if it solved it.

In general I think it's a good question on whether we want to supply such guarantees. Do you have any java code? The reason I'm asking is because this question feels like it deserves a more general Bazel JVM best practice than a specific scala one.

**Comment by johnynek** [comment]
Relying on jar order feels like an anti pattern to me.

**Comment by softprops** [comment]
> Relying on jar order feels like an anti pattern to me.

Yea that's the solution that I settled on ;). I worked around by renaming the version on the test classpath and referenced that from my test which solved the issue for me but I think everyone will have the interfaces built into their code do to so. So I'm unblocked for now but leaving this here for future reference of others are coming from sbt and hit this issue

**Comment by softprops** [comment]
That said I can think of a few more cases others may rely on. A file called `log4j.properties` is a common one. I haven't hit that yet but I'll post back here if I do.

**Comment by ittaiz** [comment]
Can we close this one for now under "won't fix"? At least until a stronger use-case will show up

**Comment by softprops** [comment]
Yes. It's notable we did actually run into this  another time with log4j.properties (and a few other misc cases). we had a third party dependency,a few actually, with log4j.properties at the root of thir jar as did we with our new bazel jar outputs. When loading these into our servlet container, we didn't get a deterministic copy resolved at runtime because the tomcat lib dir has less precedence than the classes dir, where we used to stage sbt outputs ( class files ). While this was annoying, it shouldnt have been surprising. The surprises were all just realizations bazel brought to our attention faster than with sbt. And for that we are thankful for bazel. I very much agree that order dependence should not be an expected propery of a deterministic build system. If these problems come up for others under these same circumstances it would be fair to say it's a user error.

**Comment by idancohen11** [comment]
@softprops when you say
>  I worked around by renaming the version on the test classpath and referenced that from my test
can you maybe give a bit more detail on what you mean?
As you suspected I indeed am hitting the very same issue :(

**Comment by arjantop** [comment]
We have a huge problems with this with connection to akka and typesafe config. The configuration is done in such a way that there are multiple reference.conf files in base libraries that are then extended by a reference.conf in our codebase. These two files then have to be merged in order to work correctly. This is not a problem in *_deploy.jar that merges these files (not sure exactly what is the difference at the time).
